Port of Qwen3-VL support from mainline

Thireus opened the pull request:
- convert_hf_to_gguf.py - Not touched, use llama.cpp to convert model instead
- sycl and metal support for imrope not added - See ggml-metal and ggml-sycl folders of https://github.com/ggml-org/llama.cpp/pull/16780/files
- Vulkan support for imrope not tested
- Code not tested against any GGUF (looking for testers...)

Source: https://github.com/ggml-org/llama.cpp/pull/16780

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

ranilongxi commented:
I tried to build this branch with cpu-only and CUDA but it kept failing:

CUDA
`cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DGGML_LTO=ON -DGGML_NATIVE=ON -DGGML_CUDA=ON -DGGML_CUDA_F16=ON -DGGML_CUDA_FA_ALL_QUANTS=ON -DGGML_CUDA_DMMV_F16=ON -DGGML_CUDA_FORCE_MMQ=ON -DCMAKE_CUDA_ARCHITECTURES=89 -DLLAMA_CURL=ON -DGGML_BLAS=OFF -DGGML_SCHED_MAX_COPIES=1
`

`FAILED: src/CMakeFiles/llama.dir/llama-load-tensors.cpp.o
.../llama-load-tensors.cpp:1006:13: error: conflicting declaration 'int64_t n_embd'
     int64_t n_embd = hparams.n_embd / (hparams.n_deepstack_layers + 1);
             ^~~~~~
.../llama-load-tensors.cpp:227:40: note: previous declaration as 'const int64_t n_embd'
         [[maybe_unused]] const int64_t n_embd = hparams.n_embd; \
... in expansion of macro 'LOADING_PRELUDE'
.../llama-load-tensors.cpp:1045:13: error: conflicting declaration 'int64_t n_embd'
`
CPU only
`cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Release
ninja -C build -j22`

`FAILED: src/CMakeFiles/llama.dir/llama-load-tensors.cpp.o
.../llama-load-tensors.cpp:1006:13: error: conflicting declaration 'int64_t n_embd'
.../llama-load-tensors.cpp:227:40: note: previous declaration as 'const int64_t n_embd'
... in expansion of macro 'LOADING_PRELUDE'
`

ikawrakow commented on examples/mtmd/clip.cpp:
```diff
@@ -2533,10 +2746,11 @@ struct clip_model_loader {
-            layer.k_w    = get_tensor(string_format(TN_ATTN_K,      prefix, il, "weight"));
-            layer.q_w    = get_tensor(string_format(TN_ATTN_Q,      prefix, il, "weight"));
-            layer.v_w    = get_tensor(string_format(TN_ATTN_V,      prefix, il, "weight"));
+            layer.k_w    = get_tensor(string_format(TN_ATTN_K,      prefix, il, "weight"), false);
```
Can we have `k_w, q_w, v_w`, and then also have `qkv_w`? Or not have any of those? I think, most likely what it needs to happen here is that one first tries to create `qkv_w` and, if that fails, only then one creates `k_w, q_w, v_w` (and they must be there in that case, so there is no `false` in the `get_tensor` argument list).

ikawrakow reviewed (commented):
Apart from the comments, LGTM.

But we need to get this tested by someone before merging.

Thireus commented:
Thank you @ikawrakow, I'll do my best to resolve your comments. There are still some compilation issues which I'll resolve as well.

Thireus commented:
@ikawrakow please let me know if you are happy with how I've addressed your comments:
https://github.com/ikawrakow/ik_llama.cpp/pull/883/commits/e552942129f28fac8a39bfee5897691d74f4c431
and
https://github.com/ikawrakow/ik_llama.cpp/pull/883/commits/6a24dec8b02ad3b107cd0dbab1be7d5f240467e5

MrHills-2 commented:
With llama-server and openai API text works, but sending images returns error "Failed to parse messages: Unsupported content part type: \\\"image_url\\\"\".

Thireus commented:
> With llama-server and openai API text works, but sending images returns error "Failed to parse messages: Unsupported content part type: \"image_url\"".

It's a separate issue and has been mentioned by @ikawrakow - https://github.com/ikawrakow/ik_llama.cpp/issues/792#issuecomment-3341342894

firecoperana commented:
https://github.com/ikawrakow/ik_llama.cpp/pull/901 Support of vision model in llama-server
